File: __OTHER_FRAMEWORKS__/grafana/dashboards-as-code-workshop-main/reference/part-two-golang/grpc.go
package main

import (
	"fmt"

	"github.com/grafana/grafana-foundation-sdk/go/heatmap"
	"github.com/grafana/grafana-foundation-sdk/go/logs"
	"github.com/grafana/grafana-foundation-sdk/go/prometheus"
	"github.com/grafana/grafana-foundation-sdk/go/timeseries"
	"github.com/grafana/grafana-foundation-sdk/go/units"
)

func grpcRequestsTimeseries(service Service) *timeseries.PanelBuilder {
	return timeseriesPanel().
		Title("gRPC Requests").
		Unit(units.RequestsPerSecond).
		WithTarget(prometheusQuery(fmt.Sprintf("rate(grpc_server_handled_total{service=\"%s\"}[$__rate_interval])", service.Name)).
			LegendFormat("{{ grpc_method }} – {{ grpc_code }}"),
		).
		Datasource(prometheusDatasourceRef())
}

func grpcLatenciesHeatmap(service Service) *heatmap.PanelBuilder {
	return heatmapPanel().
		Title("gRPC Requests latencies").
		WithTarget(prometheusQuery(fmt.Sprintf("sum(increase(grpc_server_handling_seconds_bucket{service=\"%s\"}[$__rate_interval])) by (le)", service.Name)).
			Format(prometheus.PromQueryFormatHeatmap),
		).
		Datasource(prometheusDatasourceRef())
}

func grpcLogsPanel(service Service) *logs.PanelBuilder {
	return logPanel().
		Title("gRPC Logs").
		WithTarget(lokiQuery(fmt.Sprintf("{service=\"%s\", source=\"grpc\", level=~\"$logs_level\"} |~ \"$logs_filter\"", service.Name)))
}

File: __OTHER_FRAMEWORKS__/grafana/dashboards-as-code-workshop-main/reference/part-two-typescript/src/grpc.ts
import * as heatmap from '@grafana/grafana-foundation-sdk/heatmap';
import * as logs from '@grafana/grafana-foundation-sdk/logs';
import * as timeseries from '@grafana/grafana-foundation-sdk/timeseries';
import * as prometheus from '@grafana/grafana-foundation-sdk/prometheus';
import * as units from '@grafana/grafana-foundation-sdk/units';
import { Service } from './catalog';
import {
    heatmapPanel,
    logPanel,
    lokiQuery,
    prometheusDatasourceRef,
    prometheusQuery,
    timeseriesPanel,
} from './common';

export const grpcRequestsTimeseries = (service: Service): timeseries.PanelBuilder => {
    return timeseriesPanel()
        .title('gRPC Requests')
        .unit(units.RequestsPerSecond)
        .withTarget(
            prometheusQuery(`rate(grpc_server_handled_total{service="${service.name}"}[$__rate_interval])`)
                .legendFormat('{{ grpc_method }} – {{ grpc_code }}')
        )
        .datasource(prometheusDatasourceRef())
    ;
};

export const grpcLatenciesHeatmap = (service: Service): heatmap.PanelBuilder => {
    return heatmapPanel()
        .title('gRPC Requests latencies')
        .withTarget(
            prometheusQuery(`sum(increase(grpc_server_handling_seconds_bucket{service="${service.name}"}[$__rate_interval])) by (le)`)
                .format(prometheus.PromQueryFormat.Heatmap)
        )
        .datasource(prometheusDatasourceRef())
    ;
};

export const grpcLogsPanel = (service: Service): logs.PanelBuilder => {
    return logPanel()
        .title('gRPC Logs')
        .withTarget(
            lokiQuery(`{service="${service.name}", source="grpc", level=~"$logs_level"} |~ "$logs_filter"`)
        )
    ;
};
